src/main/scala/corpus.scala

```diff
@@ -0,0 +1,87 @@
+package treesitter.lsp
+
+import parsley.character.*
+import parsley.debug.*
+import parsley.combinator.*
+import parsley.Parsley
+
+enum LispNode:
+  case Leaf(name: String)
+  case Nest(title: String, nodes: List[LispNode])
+
+object LispNode:
+  export parsers.parser
+
+  private object parsers:
+    def constructNode(nm: String, subnodes: List[LispNode]) =
+      subnodes match
+        case head :: next => LispNode.Nest(nm, subnodes)
+        case Nil          => LispNode.Leaf(nm)
+
+    val nodeName = some(letter <|> char('_')).map(_.mkString)
+
+    extension [A](p: => Parsley[A]) def wsp = p <* whitespaces
+
+    lazy val parser: parsley.Parsley[LispNode] = (
+      char('(') *>
+        (nodeName.wsp <~> {
+
+          val rec = many(parser)
+
+          val simple = nodeName
+            .map(LispNode.Leaf(_))
+            .map(List(_))
+
+          simple | rec
+
+        }).map(constructNode) <*
+        char(')').wsp
+    )
+  end parsers
+end LispNode
+
+case class TextCase(title: String, code: String, expected: LispNode)
+object TextCase:
+  export parsers.{textcase as parser, emptyLines}
+
+  private object parsers:
+    val titleSeparator = someUntil(char('='), endOfLine).void
+
+    val titleLine =
+      (noneOf('=') <::> someUntil(item, endOfLine)).map(_.mkString)
+
+    val title = sepEndBy1(titleLine, endOfLine).map(_.mkString("\n"))
+
+    val header =
+      titleSeparator *> title <* titleSeparator
+
+    val codeLine = (noneOf('-') <::> many(satisfy(_ != '\n')))
+      .map(_.mkString)
+
+    val code = sepEndBy(codeLine.debug("code line"), some(endOfLine))
+      .map(_.mkString("\n"))
+
+    val codeSeparator = string("---") *> manyUntil(char('-'), endOfLine)
+
+    val emptyLines =
+      sepEndBy1(many(manyUntil(whitespace, endOfLine).void), endOfLine).void
```

Comment:
There is something deeply unsettling about this...

is it not just `parsley.character.whitespaces`? `whitespace` already reads newlines, and you are `.void`ing it so you don't care about the lines themselves.

And, if it is just `whitespaces`, then I think you want the same whitespace treatment given in the lisp code above, where the "tokens" in the `lexer` object handle whitespace so the parser doesn't need to...